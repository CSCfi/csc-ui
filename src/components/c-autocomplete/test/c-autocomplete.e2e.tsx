import { E2EElement, E2EPage, newE2EPage } from '@stencil/core/testing';

describe('c-autocomplete', () => {
  let page: E2EPage;

  let autocomplete: E2EElement;

  beforeEach(async () => {
    page = await newE2EPage({ html: `<c-autocomplete></c-autocomplete>` });
    autocomplete = await page.find('c-autocomplete');
  });

  it('renders correctly', async () => {
    expect(autocomplete).toHaveClass('hydrated');
  });

  it('displays given items correctly', async () => {
    const items = [
      { name: 'Koskinen', value: 'koskinen' },
      { name: 'Kairinen', value: 'kairinen' },
    ];
    autocomplete.setProperty('items', items);

    const queryChange = await autocomplete.spyOnEvent('changeQuery');

    await page.waitForChanges();

    const input = await page.find('c-autocomplete >>> input');

    await input.click();

    await page.keyboard.type('nen');

    await page.waitForChanges();

    const renderedItems = (await page.find('c-autocomplete >>> ul'))
      .textContent;

    expect(renderedItems).toMatchSnapshot();
    expect(queryChange.events.length).toBe(3);
  });

  it('displays custom menu', async () => {
    page = await newE2EPage({
      html: `
      <c-autocomplete>
        <div slot="customMenu">
          <span>
            <b>Koskinen</b>
          </span>
          <p>Laitila</p>
        </div>
        <div slot="customMenu">
          <span>
            <b>Kairinen</b>
          </span>
          <p>Turku</p>
        </div>
      </c-autocomplete>`,
    });
    autocomplete = await page.find('c-autocomplete');
    autocomplete.setProperty('customMenu', true);

    await page.waitForChanges();

    const input = await page.find('c-autocomplete >>> input');

    await input.click();

    await page.keyboard.press(' ');

    await page.waitForChanges();

    const items = await page.evaluate(() => {
      const element = document.querySelector('c-autocomplete');

      return (
        element.shadowRoot.querySelector(
          '.c-input-menu__items slot',
        ) as HTMLSlotElement
      ).assignedNodes();
    });

    console.log(items);

    //expect(await items.isVisible()).toBe(true);

    expect(items.length).toBe(2);
  });
});
