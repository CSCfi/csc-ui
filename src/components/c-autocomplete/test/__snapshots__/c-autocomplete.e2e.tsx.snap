// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`c-autocomplete displays custom menu 1`] = `
<c-autocomplete class="hydrated">
  <mock:shadow-root>
    <div aria-atomic="true" aria-live="polite" class="visuallyhidden" id="announce-7e2c1a55-eccc-4d6b-94e0-0f033877e7b5"></div>
    <c-input class="hydrated" input-id="input_">
      <!---->
      <div aria-atomic="true" aria-live="polite" class="visuallyhidden" id="announce-input_"></div>
      <div class="c-input c-input--text">
        <div class="c-input__control">
          <div class="c-input__slot">
            <fieldset aria-hidden="true">
              <legend class="active" style="--c-legend-width: 0px;">
                <span class="notranslate"></span>
              </legend>
            </fieldset>
            <div class="c-input__field" style="--c-label-position: 0px;">
              <slot name="pre" slot="pre"></slot>
              <div class="c-input__content">
                <div class="c-input-menu__input">
                  <input aria-activedescendant="option--7e2c1a55-eccc-4d6b-94e0-0f033877e7b5-null" aria-autocomplete="list" aria-expanded="false" aria-owns="results_7e2c1a55-eccc-4d6b-94e0-0f033877e7b5" autocomplete="off" class="c-input__input" placeholder="" role="combobox" type="text">
                </div>
                <div class="c-input-menu__item-wrapper"></div>
                <svg class="c-input-menu__chevron" viewBox="0 0 24 24">
                  <path d="M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z"></path>
                </svg>
              </div>
              <slot name="post" slot="post"></slot>
            </div>
          </div>
          <div class="active c-input__details">
            <div class="c-input__message c-input__message--hint">
              <span id="hint-input_">
                <span class="visuallyhidden">
                  Hint:
                </span>
              </span>
            </div>
          </div>
        </div>
      </div>
    </c-input>
    <style>
      :host{--csc-border-radius:4px;--csc-dark-grey:rgb(89, 89, 89);--csc-error:#e71d32;--csc-font-family:'museo-sans', sans-serif;--csc-light-grey:rgb(223, 225, 227);--csc-light-grey-blue:rgb(204, 244, 240);--csc-lightest-grey:rgba(223, 225, 227, 0.5);--csc-link:#025B97;--csc-mid-grey:rgb(140, 140, 140);--csc-primary:rgb(0, 103, 120);--csc-primary-ghost:rgba(0, 103, 120, 0.15);--csc-primary-ghost-hover:rgba(0, 103, 120, 0.25);--csc-primary-hover:rgb(80, 151, 141);--csc-primary-text-hover:rgba(0, 103, 120, 0.15);--csc-success:#51a808;--csc-warning:#ff5800}html{box-sizing:border-box}*,*:before,*:after{box-sizing:inherit}.md-ripple{display:block;position:absolute;pointer-events:none;border-radius:50%;transform:scale(0);background:currentColor;opacity:0.3}.md-ripple.animate{animation:mdRipple 0.7s backwards linear}:host{font-family:var(--csc-font-family)}@keyframes mdRipple{100%{opacity:0;transform:scale(2.5)}}::-ms-reveal{display:none}:host{display:block;cursor:text}:host([disabled]) ::slotted(input),:host([disabled]) ::slotted(textarea){color:rgba(0, 0, 0, 0.5)}.visuallyhidden{border:0;clip:rect(0 0 0 0);height:1px;margin:-1px;overflow:hidden;padding:0;position:absolute;white-space:nowrap;width:1px}.c-input{--c-input-border-width:1px;align-items:flex-start;display:flex;flex:1 1 auto;font-family:"museo-sans", sans-serif;font-size:16px;letter-spacing:normal;max-width:100%;text-align:left;border-radius:4px}.c-input input,.c-input select,.c-input textarea{margin:0;border:0;padding:0;display:inline-block;vertical-align:middle;white-space:normal;background:none;line-height:1;font-size:16px;font-family:"museo-sans", sans-serif}.c-input input,.c-input textarea{-webkit-box-sizing:content-box;-moz-box-sizing:content-box;box-sizing:content-box}.c-input input[type=checkbox],.c-input input[type=radio],.c-input select{-webkit-box-sizing:border-box;-moz-box-sizing:border-box;box-sizing:border-box}.c-input input[type=search]{-webkit-appearance:textfield;-webkit-box-sizing:content-box;box-sizing:content-box}.c-input ::-webkit-search-decoration{display:none}.c-input textarea{vertical-align:top;overflow:auto}.c-input input{max-height:32px;padding:8px 0 8px}.c-input textarea{margin-top:4px;min-height:44px;padding:8px 12px 8px 0;resize:vertical}.c-input input,.c-input textarea{background-color:transparent;border:none;color:rgba(0, 0, 0, 0.87);flex:1 1 auto;font-family:"museo-sans", sans-serif;font-size:16px;line-height:20px;max-width:100%;min-width:0;width:100%}.c-input input:focus,.c-input input:active,.c-input textarea:focus,.c-input textarea:active{outline:none}.c-input .c-input__content{align-items:center;display:flex;width:100%}.c-input__control{border-radius:inherit;color:currentColor;display:flex;flex-direction:column;flex-grow:1;flex-wrap:wrap;gap:8px;height:auto;min-width:0;width:100%}.c-input__slot{align-items:stretch;background:transparent;border-radius:inherit;color:inherit;cursor:text;display:flex;min-height:44px;padding:0 12px;position:relative;transition:0.3s cubic-bezier(0.25, 0.8, 0.5, 1);width:100%}.c-input__slot .click-trap{position:absolute;top:0;left:0;width:100%;height:100%;background-color:rgba(128, 255, 0, 0.295)}.c-input__field{align-items:center;display:flex;flex:1 1 auto;gap:8px;position:relative}.c-input__field input,.c-input__field textarea{flex:1;outline:0}.c-input__field label{font-size:16px;height:20px;left:0px;letter-spacing:normal;line-height:20px;max-width:90%;min-height:8px;overflow:hidden;pointer-events:none;position:absolute;right:auto;text-overflow:ellipsis;top:12px;transform-origin:top left;transform:translateX(var(--c-label-position)) translateY(0) scale(1);transition:0.3s cubic-bezier(0.25, 0.8, 0.5, 1);white-space:nowrap}.c-input__field label.active{transform:translateY(-18px) scale(0.75)}.c-input__field label span{color:var(--csc-error)}.c-input__details{line-height:1;padding:0 12px}.c-input__details.active .c-input__message{opacity:1;transform:translateY(0px)}.c-input__message{font-size:12px;min-height:16px;opacity:0;transform:translateY(-4px);transition:all 0.2s cubic-bezier(0.25, 0.8, 0.5, 1)}.c-input__message--hint{color:var(--csc-mid-grey)}.c-input__message--error{color:var(--csc-error)}.c-input__message span{align-items:flex-start;color:currentColor;display:flex;gap:4px;min-height:16px}.c-input__message svg{fill:currentColor;height:16px;width:16px;position:relative;top:-2px}.c-input--select .c-input__slot{cursor:pointer}.c-input fieldset{border-collapse:collapse;border-color:currentColor;border-radius:inherit;border-style:solid;border-width:var(--c-input-border-width);bottom:0;left:0;margin:0;padding-left:8px;padding:0;pointer-events:none;position:absolute;right:0;top:0;transition-duration:0.15s;transition-property:color;transition-timing-function:cubic-bezier(0.25, 0.8, 0.25, 1)}.c-input fieldset legend{float:none;line-height:11px;margin-left:7px;padding:0;text-align:left;transition:width 0.3s cubic-bezier(0.25, 0.8, 0.5, 1);width:0}.c-input fieldset legend.active{width:var(--c-legend-width)}.c-input--shadow .c-input__slot{background:#fff;box-shadow:rgba(0, 0, 0, 0.15) 0px 5px 15px 0px}.c-input--shadow .c-input__slot:focus-within{outline:2px var(--csc-primary) solid}.c-input--disabled input,.c-input--disabled textarea,.c-input--disabled fieldset,.c-input--disabled label{--csc-error:rgba(0, 0, 0, 0.5);color:rgba(0, 0, 0, 0.5)}.c-input--textarea .c-input__field{margin-right:-12px}.c-input ::placeholder{color:rgba(0, 0, 0, 0.46);opacity:1}.c-input:focus-within,.c-input:not(.c-input--disabled):active{--c-input-border-width:2px;color:var(--csc-primary)}.c-input:focus-within label,.c-input:not(.c-input--disabled):active label{transform:translateX(0px) translateY(-18px) scale(0.75)}.c-input:focus-within legend,.c-input:not(.c-input--disabled):active legend{width:var(--c-legend-width)}.c-input--error{color:var(--csc-error)}.c-input--error:focus-within{color:var(--csc-error)}
    </style>
  </mock:shadow-root>
  <div slot="customMenu">
    <span>
      <b>
        Koskinen
      </b>
    </span>
    <p>
      Laitila
    </p>
  </div>
  <div slot="customMenu">
    <span>
      <b>
        Kairinen
      </b>
    </span>
    <p>
      Turku
    </p>
  </div>
</c-autocomplete>
`;

exports[`c-autocomplete displays given items correctly 1`] = `"KoskinenKairinen"`;
